flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/collect/CollectResultFetcher.java

```diff
@@ -0,0 +1,343 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.api.operators.collect;
+
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.api.common.JobExecutionResult;
+import org.apache.flink.api.common.JobStatus;
+import org.apache.flink.api.common.accumulators.SerializedListAccumulator;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.api.common.typeutils.base.array.BytePrimitiveArraySerializer;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.core.execution.JobClient;
+import org.apache.flink.runtime.jobgraph.OperatorID;
+import org.apache.flink.runtime.operators.coordination.CoordinationRequestGateway;
+import org.apache.flink.util.Preconditions;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * A fetcher which fetches query results from sink and provides exactly-once semantics.
+ */
+public class CollectResultFetcher<T> {
+
+	private static final int DEFAULT_RETRY_MILLIS = 100;
+	private static final long DEFAULT_ACCUMULATOR_GET_MILLIS = 10000;
+
+	private static final Logger LOG = LoggerFactory.getLogger(CollectResultFetcher.class);
+
+	private final CompletableFuture<OperatorID> operatorIdFuture;
+	private final String accumulatorName;
+	private final int retryMillis;
+
+	private ResultBuffer buffer;
+
+	private JobClient jobClient;
+	private boolean terminated;
+	private boolean closed;
+
+	public CollectResultFetcher(
+			CompletableFuture<OperatorID> operatorIdFuture,
+			TypeSerializer<T> serializer,
+			String accumulatorName) {
+		this(
+			operatorIdFuture,
+			serializer,
+			accumulatorName,
+			DEFAULT_RETRY_MILLIS);
+	}
+
+	@VisibleForTesting
+	public CollectResultFetcher(
+			CompletableFuture<OperatorID> operatorIdFuture,
+			TypeSerializer<T> serializer,
+			String accumulatorName,
+			int retryMillis) {
+		this.operatorIdFuture = operatorIdFuture;
+		this.accumulatorName = accumulatorName;
+		this.retryMillis = retryMillis;
+
+		this.buffer = new ResultBuffer(serializer);
+
+		this.terminated = false;
+	}
+
+	public void setJobClient(JobClient jobClient) {
+		Preconditions.checkArgument(
+			jobClient instanceof CoordinationRequestGateway,
+			"Job client must be a CoordinationRequestGateway. This is a bug.");
+		this.jobClient = jobClient;
```

Comment:
use another field to save `CoordinationRequestGateway`, can save some casting in the future